**This notebook is an exercise in the [Python](https://www.kaggle.com/learn/python) course.  You can reference the tutorial at [this link](https://www.kaggle.com/colinmorris/functions-and-getting-help).**

---


Functions are powerful. Try writing some yourself.

As before, don't forget to run the setup code below before jumping into question 1.

関数は強力だ。自分でいくつか書いてみてください。

前回と同様、問題1に入る前に以下のセットアップコードを実行することを忘れないでください。}

In [1]:
# SETUP. You don't need to worry for now about what this code does or how it works.
from learntools.core import binder; binder.bind(globals())
from learntools.python.ex2 import *
print('Setup complete.')

Setup complete.


# 1.

Complete the body of the following function according to its docstring.

HINT: Python has a built-in function `round`.


# 1.

docstring に従って以下の関数本体を完成させなさい。

ヒント：Pythonには組み込み関数 `round` がある。

In [2]:
def round_to_two_places(num):
    """Return the given number rounded to two decimal places. 
    return round(num, 2)
    >>> 
    3.14
    """
    return round(num, 2)
    # Replace this body with your own code.
    # ("pass" is a keyword that does literally nothing. We used it as a placeholder
    # because after we begin a code block, Python requires at least one line of code)
    pass




# Check your answer
q1.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct</span>

In [3]:
# Uncomment the following for a hint
#q1.hint()
# Or uncomment the following to peek at the solution
q1.solution()

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> 
```python
return round(num, 2)
```

# 2.
The help for `round` says that `ndigits` (the second argument) may be negative.
What do you think will happen when it is? Try some examples in the following cell.

# 2.
round`のヘルプによると、`ndigits`（第二引数）は負になる可能性がある。
その場合どうなると思いますか？次のセルでいくつかの例を試してみてください。

In [4]:
# Put your test code here

Can you think of a case where this would be useful?  Once you're ready, run the code cell below to see the answer and to receive credit for completing the problem.

In [5]:
# Check your answer (Run this code cell to receive credit!)
q2.solution()

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> As you've seen, `ndigits=-1` rounds to the nearest 10, `ndigits=-2` rounds to the nearest 100 and so on. Where might this be useful? Suppose we're dealing with large numbers:

> The area of Finland is 338,424 km²  
> The area of Greenland is 2,166,086 km²

We probably don't care whether it's really 338,424, or 338,425, or 338,177. All those digits of accuracy are just distracting. We can chop them off by calling `round()` with `ndigits=-3`:

> The area of Finland is 338,000 km²  
> The area of Greenland is 2,166,000 km²

(We'll talk about how we would get the commas later when we talk about string formatting :))


解答 おわかりのように、ndigits=-1 は 10 の位を四捨五入し、ndigits=-2 は 100 の位を四捨五入する。これはどのような場面で役に立つだろうか？大きな数を扱っているとしよう：

フィンランドの面積は338,424 km² である。
グリーンランドの面積は2,166,086 km² です。

それが本当に338,424km²なのか、338,425km²なのか、338,177km²なのか、私たちはおそらく気にしない。桁の精度は邪魔なだけだ。ndigits=-3でround()を呼び出すことで、桁数を減らすことができる：

フィンランドの面積は338,000km²である。
グリーンランドの面積は2,166,000 km² です。

(どのようにカンマを取るかについては、後で文字列のフォーマットについて話すときに話そう :))

# 3.

In the previous exercise, the candy-sharing friends Alice, Bob and Carol tried to split candies evenly. For the sake of their friendship, any candies left over would be smashed. For example, if they collectively bring home 91 candies, they'll take 30 each and smash 1.

Below is a simple function that will calculate the number of candies to smash for *any* number of total candies.

Modify it so that it optionally takes a second argument representing the number of friends the candies are being split between. If no second argument is provided, it should assume 3 friends, as before.

Update the docstring to reflect this new behaviour.

# 3.

前の練習では、キャンディーを分け合う友人アリス、ボブ、キャロルは、キャンディーを均等に分けようとした。友情のために、残ったキャンディーは潰してしまう。例えば、91個のキャンディを持って帰ったら、30個ずつ取って、1個をつぶす。

以下は、キャンディーの総数が*任意の*数である場合に、粉砕するキャンディーの数を計算する簡単な関数である。

この関数を修正して、キャンディを分ける友達の数を表す第二引数を取るようにする。第2引数が与えられない場合は、以前と同様に3人の友達と仮定する。

この新しい動作を反映するようにdocstringを更新してください。

In [6]:


def to_smash(total_candies):
    """Return the number of leftover candies that must be smashed after distributing
    the given number of candies evenly between 3 friends.
    
    >>> to_smash(91)
    1
    """
    return total_candies % 3

def to_smash(total_candies, n_friends=3):
    return total_candies % n_friends

# Check your answer
q3.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct</span>

In [7]:
#q3.hint()

In [8]:
q3.solution()

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> 
```python
def to_smash(total_candies, n_friends=3):
    return total_candies % n_friends
```

# 4. (Optional)

It may not be fun, but reading and understanding error messages will be an important part of your Python career.

Each code cell below contains some commented buggy code. For each cell...

1. Read the code and predict what you think will happen when it's run.
2. Then uncomment the code and run it to see what happens. (**Tip**: In the kernel editor, you can highlight several lines and press `ctrl`+`/` to toggle commenting.)
3. Fix the code (so that it accomplishes its intended purpose without throwing an exception)

楽しくはないかもしれませんが、エラーメッセージを読んで理解することはあなたのPythonキャリアにとって重要な部分となるでしょう。

以下の各コードセルには、コメントされたバグのあるコードが含まれています。それぞれのセルについて...

1. コードを読んで、それが実行されたときに何が起こると思うか予想しなさい。
2. 次に、コードのコメントを外して実行し、何が起こるか見てください。(**ヒント**： カーネルエディターでは、いくつかの行をハイライトし、`ctrl`+`/` を押すとコメントを切り替えることができます)。
3. コードを修正する（例外を投げずに本来の目的を達成できるように）。

<!-- TODO: should this be autochecked? Delta is probably pretty small. -->

In [9]:
round_to_two_places(9.9999)

10.0

In [10]:
x = -10
y = 5
# Which of the two variables above has the smallest absolute value?
def least_difference(x,y):
    diff1 = abs(x)
    diff2 = abs(y)
    return min(diff1, diff2)

In [11]:
def mult_by_five(x):
    return 5 * x

def call(fn, arg):
    """Call fn on arg"""
    return fn(arg)


# Keep Going

Nice job with the code. Next up, you'll learn about *conditionals*, which you'll need to **[write interesting programs](https://www.kaggle.com/colinmorris/booleans-and-conditionals)**. 

---




*Have questions or comments? Visit the [course discussion forum](https://www.kaggle.com/learn/python/discussion) to chat with other learners.*